src/runtime/uv/tcp.cpp

```diff
@@ -0,0 +1,599 @@
+
+    event_loop_lock(&global_ev);
+    int result = uv_accept((uv_stream_t*)tcp_socket->m_uv_tcp, (uv_stream_t*)client_socket->m_uv_tcp);
+    event_loop_unlock(&global_ev);
+
+    if (result < 0 && result != UV_EAGAIN) {
+        lean_dec(client);
+        resolve_promise_with_status(promise, result);
+    } else if (result >= 0) {
+        resolve_promise(promise, mk_ok_except(client));
+    } else {
+        // The event loop owns the object. It will be released in the listen
+        lean_inc(socket);
+
+        lean_inc(promise);
+        tcp_socket->m_promise_accept = promise;
+        tcp_socket->m_client = client;
+    }
+
+    return lean_io_result_mk_ok(promise);
+}
+
+/* Std.Internal.UV.TCP.Socket.shutdown (socket : @& Socket) : IO (IO.Promise (Except IO.Error Unit)) */
+extern "C" LEAN_EXPORT lean_obj_res lean_uv_tcp_shutdown(b_obj_arg socket) {
+    lean_uv_tcp_socket_object * tcp_socket = lean_to_uv_tcp_socket(socket);
+
+    if (tcp_socket->m_promise_shutdown != nullptr) {
+        return lean_io_result_mk_error(lean_decode_uv_error(UV_EALREADY, mk_string("shutdown already in progress")));
+    }
+
+    lean_object * promise = create_promise();
+    tcp_socket->m_promise_shutdown = promise;
+    lean_inc(promise);
+
+    event_loop_lock(&global_ev);
+
+    uv_shutdown_t* shutdown_req = (uv_shutdown_t*)malloc(sizeof(uv_shutdown_t));
+    shutdown_req->data = (void*)socket;
+
+    lean_inc(socket);
+
+    int result = uv_shutdown(shutdown_req, (uv_stream_t*)tcp_socket->m_uv_tcp, [](uv_shutdown_t* req, int status) {
+        lean_uv_tcp_socket_object * tcp_socket = lean_to_uv_tcp_socket((lean_object*)req->data);
+
+        if (status < 0) {
+            resolve_promise_with_status(tcp_socket->m_promise_shutdown, status);
+        } else {
+            resolve_promise(tcp_socket->m_promise_shutdown, mk_ok_except(lean_box(0)));
+        }
+
+        lean_dec(tcp_socket->m_promise_shutdown);
+
+        tcp_socket->m_promise_shutdown = nullptr;
+
+        lean_dec((lean_object*)req->data);
+        free(req);
+    });
+
+    event_loop_unlock(&global_ev);
+
+    if (result < 0) {
+        free(shutdown_req);
+        tcp_socket->m_promise_shutdown = nullptr;
+        return lean_io_result_mk_error(lean_decode_uv_error(result, nullptr));
+    }
+
+    return lean_io_result_mk_ok(promise);
+}
+
+/* Std.Internal.UV.TCP.Socket.getpeername (socket : @& Socket) : IO SocketAddress */
+extern "C" LEAN_EXPORT lean_obj_res lean_uv_tcp_getpeername(b_obj_arg socket) {
+    lean_uv_tcp_socket_object *tcp_socket = lean_to_uv_tcp_socket(socket);
+
+    struct sockaddr addr_storage;
+    int addr_len = sizeof(addr_storage);
+
+    event_loop_lock(&global_ev);
+    int result = uv_tcp_getpeername(tcp_socket->m_uv_tcp, (struct sockaddr*)&addr_storage, &addr_len);
+    event_loop_unlock(&global_ev);
+
+    if (result < 0) {
+        return io_result_mk_error(uv_strerror(result));
```

Comment:
Why are we using `uv_strerror` here when all the other functions use `lean_decode_uv_error`?